File: ui/admin-portal/src/app/model/job.ts
/*
 * Copyright = c) 2022 Talent Beyond Boundarie.
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */
import {SavedList, ShortSavedList} from "./saved-list";
import {User} from "./user";
import {Employer, ShortPartner} from "./partner";
import {SavedSearch} from "./saved-search";
import {Router} from "@angular/router";
import {Location} from "@angular/common";
import {getExternalHref} from "../util/url";
import {JobOppIntake} from "./job-opp-intake";
import {isCandidateOpportunity, SearchOpportunityRequest} from "./candidate-opportunity";
import {Opportunity, OpportunityProgressParams} from "./opportunity";
import {Country} from "./country";
import {getOrdinal} from "../util/enum";

export function isJob(opp: Opportunity): opp is Job {
  return !isCandidateOpportunity(opp);
}

export interface JobNameAndId {
  id: number,
  name: string;
}

export interface ShortJob extends JobNameAndId {
  country?: Country;
  submissionList?: ShortSavedList;
  jobCreator?: ShortPartner;
}

export interface Job extends Opportunity {
  hiringCommitment: string;
  opportunityScore: string;
  contactUser: User;
  country: Country;
  employerEntity: Employer;
  evergreen: boolean;
  exclusionList: SavedList;
  jobSummary: string;
  publishedBy: User;
  publishedDate: Date;
  jobCreator: ShortPartner;
  skipCandidateSearch: boolean;
  stage: JobOpportunityStage;
  starringUsers: User[];
  submissionDueDate: Date;
  submissionList: SavedList;
  suggestedList: SavedList;
  suggestedSearches: SavedSearch[];
  jobOppIntake: JobOppIntake;
}

export function getJobExternalHref(router: Router, location: Location, job: Job): string {
  return getExternalHref(router, location, ['job', job.id]);
}

export type JobDocType = "jd" | "joi" | "interview" | "mou";

/**
 * Note that the string values of this enum MUST match the actual stage names for job
 * opportunities on Salesforce.
 * <p/>
 * See https://docs.google.com/document/d/1B6DmpYaONV_yNmyAqL76cu0TUQcpNgKtOmKELCkpRoc/edit#heading=h.qx7je1tuwoqv
 */
export enum JobOpportunityStage {
  prospect = "0. Prospect",
  briefing = "1. Briefing",
  pitching = "2. Pitching",
  mou = "3. MOU",
  identifyingRoles = "4. Identifying roles",
  candidateSearch = "5. Candidate search",
  visaEligibility = "6. Visa eligibility",
  cvPreparation = "7. CV preparation",
  cvReview = "8. CV review",
  recruitmentProcess = "9. Recruitment process",
  jobOffer = "10. Job offer",
  training = "11. Training",
  visaPreparation = "12. Visa preparation",
  postHireEngagement = "13. Post hire engagement",
  hiringCompleted = "14. Closed won. Hiring completed",
  ineligibleEmployer = "Closed. Ineligible employer",
  ineligibleOccupation = "Closed. Ineligible occupation",
  ineligibleRegion = "Closed. Ineligible region",
  noInterest = "Closed. No interest",
  noJobOffer = "Closed. No job offer",
  noPrPathway = "Closed. No PR pathway",
  noSuitableCandidates = "Closed. No suitable candidates",
  noVisa = "Closed. No visa",
  tooExpensive = "Closed. Too expensive",
  tooHighWage = "Closed. Too high wage",
  tooLong = "Closed. Too long",
  mouIssue = "Closed. MOU issue",
  trainingNotCompleted = "Closed. Training not completed",
}

/**
 * Adds extra job opportunity specific fields to standard SearchOpportunityRequest
 */
export class SearchJobRequest extends SearchOpportunityRequest {
  starred?: boolean;
  jobNameAndIdOnly?: boolean;
}

export interface UpdateJobRequest extends OpportunityProgressParams {
  contactUserId?: number;
  evergreen?: boolean;
  roleName?: string;
  sfId?: string;
  jobName?: string;
  sfJoblink?: string;
  skipCandidateSearch?: boolean;
  submissionDueDate?: Date;
  jobToCopyId?: number;
}


export function isJobOppStageGreaterThanOrEqualTo(selectedOppStageKey: string, desiredStageKey: string) {
  let oppOrdinal: number = getOrdinal(JobOpportunityStage, selectedOppStageKey);
  let desiredOrdinal: number = getOrdinal(JobOpportunityStage, desiredStageKey);
  return oppOrdinal >= desiredOrdinal;
}

File: server/src/main/java/org/tbbtalent/server/model/db/Job.java
/*
 * Copyright (c) 2022 Talent Beyond Boundaries.
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */

package org.tbbtalent.server.model.db;

import java.time.OffsetDateTime;
import javax.annotation.Nullable;
import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.JoinColumn;
import javax.persistence.OneToOne;
import javax.persistence.SequenceGenerator;
import javax.persistence.Table;
import lombok.Getter;
import lombok.Setter;
import lombok.ToString;

/**
 * A job represents a Job Opportunity provided by an employer in a destination location.
 * <p/>
 * This data is populated from the corresponding Salesforce Employer Job Opportunity.
 * @author John Cameron
 */
@Getter
@Setter
@ToString
@Entity
@Table(name = "job")
@SequenceGenerator(name = "seq_gen", sequenceName = "job_id_seq", allocationSize = 1)
public class Job extends AbstractDomainObject<Long> {

    /**
     * This is the corresponding job opportunity data copied (ie cached) from Salesforce.
     */
    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "sf_job_opp_id")
    private SalesforceJobOpp sfJobOpp;

    /**
     * Date that submission of candidates to employer is due.
     */
    @Nullable
    private OffsetDateTime submissionDueDate;

    /**
     * This is the official list of candidates which will be submitted to the employer for
     * their consideration.
     * <p/>
     * SubmissionList should be a registeredJob associated with sfJobOpp
     */
    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "submission_list_id")
    private SavedList submissionList;

    public String getCountry() {
        return sfJobOpp == null ? null :sfJobOpp.getCountry();
    }

    public String getEmployer() {
        return sfJobOpp == null ? null :sfJobOpp.getEmployer();
    }

    public String getName() {
        return sfJobOpp == null ? null :sfJobOpp.getName();
    }

    public JobOpportunityStage getStage() {
        return sfJobOpp == null ? null :sfJobOpp.getStage();
    }

    public boolean isClosed() {
        return sfJobOpp == null ? false :sfJobOpp.isClosed();
    }
}
